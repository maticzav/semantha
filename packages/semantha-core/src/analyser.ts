import { GithubCommit } from './github'
import { constants, SemanthaVersion } from './constants'
import { filterMap } from './utils'

export interface Package {
  name: string
  version: string
  dependencies: { [dependency: string]: string }
  devDependencies: { [dependency: string]: string }
}

export interface Workspace {
  path: string
  pkg: Package
}

export interface SemanthaRule {
  regex: RegExp
  release: SemanthaVersion
}

export interface SemanthaRelease {
  version: SemanthaVersion
  workspace: Workspace
  commits: GithubCommit[]
}

/**
 *
 * Analyze commits analyzes commits and their commit messages to find
 * the new version of workspace packages.
 *
 * @param workspaces
 * @param commits
 */
export function analyzeCommits(
  workspaces: Workspace[],
  commits: GithubCommit[],
  rules: SemanthaRule[],
): SemanthaRelease[] {
  /** Bind commits to workspace */
  const workspacesWithImpactingCommits = workspaces.map(workspace =>
    findWorkspaceImpactingCommits(workspace),
  )

  /** Calculate next version from commit messages */
  const workspaceVersionsFromCommitMessages = workspacesWithImpactingCommits.map(
    workspaceWitnImpactingCommits =>
      analyseCommitMessages(workspaceWitnImpactingCommits),
  )

  /** Accounts for changes in peer dependencies */
  const workspaceVersionsFromDependencies = workspaceVersionsFromCommitMessages.map(
    releaseFromCommitMessages =>
      analyseDependecyVersioning(
        releaseFromCommitMessages,
        workspaceVersionsFromCommitMessages,
      ),
  )

  return workspaceVersionsFromDependencies

  /**
   * Helper functions
   */
  /**
   *
   * Find commits which impacted files in the workspace path.
   *
   * @param workspace
   */
  function findWorkspaceImpactingCommits(
    workspace: Workspace,
  ): { workspace: Workspace; commits: GithubCommit[] } {
    const impactingCommits = commits.filter(commit =>
      commit.files.some(file => file.filename.startsWith(workspace.path)),
    )

    return {
      workspace: workspace,
      commits: impactingCommits,
    }
  }

  /**
   * Analyses commits messages to determine next version of the package.
   */
  function analyseCommitMessages({
    workspace,
    commits,
  }: {
    workspace: Workspace
    commits: GithubCommit[]
  }): SemanthaRelease {
    const version = rules.reduce((acc, rule) => {
      if (
        commits.some(commit => rule.regex.test(commit.message)) &&
        rule.release > acc
      ) {
        return rule.release
      } else {
        return acc
      }
    }, constants.IGNORE)

    return {
      version: version,
      workspace: workspace,
      commits: commits,
    }
  }

  /**
   *
   * Determines the actual version based on dependencies.
   *
   * @param releases
   */
  function analyseDependecyVersioning(
    release: SemanthaRelease,
    releases: SemanthaRelease[],
    path: string[] = [],
  ): SemanthaRelease {
    /* Merge all dependencies of a workspace */
    const dependencies: { [name: string]: string } = {
      ...release.workspace.pkg.dependencies,
      ...release.workspace.pkg.devDependencies,
    }

    /* Find related releases */
    const peerReleases = filterMap(
      dependency =>
        releases.find(release => release.workspace.pkg.name === dependency),
      Object.keys(dependencies),
    )

    /* Analyse releases */
    const version = peerReleases.reduce((acc, release) => {
      /* Path doesn't yet include the observed workspace. */
      if (!path.includes(release.workspace.pkg.name)) {
        /* Calculate dependency version */
        const dependency = analyseDependecyVersioning(release, releases, [
          ...path,
          release.workspace.pkg.name,
        ])

        /* Update versioning if it impacts the current package */
        if (dependency.version > acc) {
          return dependency.version
        } else {
          return acc
        }
      } else {
        /* Break circular dependencies. */
        return acc
      }
    }, release.version)

    return {
      version: version,
      workspace: release.workspace,
      commits: release.commits,
    }
  }
}
