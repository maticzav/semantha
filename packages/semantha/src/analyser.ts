import { Workspace } from './workspaces'
import { GithubCommit } from './github'

export interface Release {
  status: 'major' | 'minor' | 'patch' | 'ignore'
  workspace: Workspace
  commits: GithubCommit[]
}

/**
 *
 * Analyze commits analyzes commits and their commit messages to find
 * the new version of workspace packages.
 *
 * @param workspaces
 * @param commits
 */
export function analyzeCommits(
  workspaces: Workspace[],
  commits: GithubCommit[],
): Release[] {
  /** Bind commits to workspace */
  const workspacesWithImpactingCommits = workspaces.map(workspace =>
    findWorkspaceImpactingCommits(workspace),
  )

  /** Calculate next version from commit messages */
  const releasesFromCommitMessages = workspacesWithImpactingCommits.map(
    workspaceWitnImpactingCommits =>
      analyseCommitMessages(workspaceWitnImpactingCommits),
  )

  /** Accounts for changes in peer dependencies */
  const releasesFromDependencies = releasesFromCommitMessages.map(
    releaseFromCommitMessages =>
      analyseDependecyVersioning(
        releaseFromCommitMessages,
        releasesFromCommitMessages,
      ),
  )

  return releasesFromDependencies

  /**
   * Helper functions
   */
  /**
   *
   * Find commits which impacted files in the workspace path.
   *
   * @param workspace
   */
  function findWorkspaceImpactingCommits(
    workspace: Workspace,
  ): { workspace: Workspace; commits: GithubCommit[] } {
    const impactingCommits = commits.filter(commit =>
      commit.files.some(file => file.filename.startsWith(workspace.path)),
    )

    return {
      workspace: workspace,
      commits: impactingCommits,
    }
  }

  /**
   * Analyses commits messages to determine next version of the package.
   */
  function analyseCommitMessages({
    workspace,
    commits,
  }: {
    workspace: Workspace
    commits: GithubCommit[]
  }): Release {
    const status = analyseCommits(commits)

    return {
      status: status,
      workspace: workspace,
      commits: commits,
    }
  }

  /**
   *
   * Determines the actual version based on dependencies.
   *
   * @param releases
   */
  function analyseDependecyVersioning(
    release: Release,
    releases: Release[],
  ): Release {
    return {
      status: 'major',
      workspace: release.workspace,
      commits: release.commits,
    }
  }
}

/** Taken from semantic-release/commit-analyser */
const releaseRules = [
  { breaking: true, release: 'major' },
  { revert: true, release: 'patch' },
  // Angular
  { type: 'feat', release: 'minor' },
  { type: 'fix', release: 'patch' },
  { type: 'perf', release: 'patch' },
  // Atom
  { emoji: ':racehorse:', release: 'patch' },
  { emoji: ':bug:', release: 'patch' },
  { emoji: ':penguin:', release: 'patch' },
  { emoji: ':apple:', release: 'patch' },
  { emoji: ':checkered_flag:', release: 'patch' },
  // Ember
  { tag: 'BUGFIX', release: 'patch' },
  { tag: 'FEATURE', release: 'minor' },
  { tag: 'SECURITY', release: 'patch' },
  // ESLint
  { tag: 'Breaking', release: 'major' },
  { tag: 'Fix', release: 'patch' },
  { tag: 'Update', release: 'minor' },
  { tag: 'New', release: 'minor' },
  // Express
  { component: 'perf', release: 'patch' },
  { component: 'deps', release: 'patch' },
  // JSHint
  { type: 'FEAT', release: 'minor' },
  { type: 'FIX', release: 'patch' },
]

/**
 *
 * Analyses commits using commit messages.
 *
 * @param commits
 */
function analyseCommits(
  commits: GithubCommit[],
): 'major' | 'minor' | 'patch' | 'ignore' {
  return 'major'
}
