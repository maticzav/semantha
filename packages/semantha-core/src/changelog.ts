import { SemanthaRelease } from './analyser'

export function generateChangelog(release: SemanthaRelease): string {
  return ''
}
