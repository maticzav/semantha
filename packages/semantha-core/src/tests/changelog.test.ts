import { SemanthaRelease } from '../analyser'
import { generateChangelog } from '../changelog'

describe('changelog', () => {
  test('generateChangelog generates changelog as expected', async () => {
    const release: SemanthaRelease = {
      commits: [
        {
          body: '',
          files: [{ filename: '/packages/package-c/package.json' }],
          message: 'feat: Minor change package-c',
          sha: '',
        },
      ],
      version: 2,
      workspace: {
        path: '/packages/package-c',
        pkg: {
          _id: 'package-c@',
          dependencies: { irrelavant: '1.0.0', 'package-d': '1.0.0' },
          devDependencies: { irrelavantDev: '1.0.0' },
          name: 'package-c',
          readme: 'ERROR: No README data found!',
          version: '1.0.0',
        },
      },
    }

    expect(generateChangelog(release)).toMatchSnapshot()
  })
})
