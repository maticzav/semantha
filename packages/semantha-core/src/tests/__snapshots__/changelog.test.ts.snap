// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changelog generateChangelog generates changelog as expected 1`] = `""`;
